src/Native/Unix/System.Security.Cryptography.Native/pal_bio.cpp

```diff
@@ -52,3 +52,78 @@ extern "C" int32_t CryptoNative_BioCtrlPending(BIO* bio)
     assert(result <= INT32_MAX);
     return static_cast<int32_t>(result);
 }
+
+extern "C" void CryptoNative_BioSetAppData(BIO* bio, void* data)
+{
+    BIO_set_ex_data(bio, 0, data);
+}
+
+extern "C" void* CryptoNative_BioGetAppData(BIO* bio)
+{
+    return BIO_get_ex_data(bio, 0);
```

Comment:
Right now every callback needs to marshal back in to get this value.  Is it reasonable to change the signature for the managed callbacks so that this gets passed directly in, and do this call in native code before invoking the callback?

It might be a big pain and not really worth it; but it's a thought.